File: Tedesco.Core/Model/Pitch/Interval.cs
using System;

namespace Tedesco
{
	public sealed class Interval : IEquatable<Interval>, IComparable<Interval>, IComparable
	{
        private readonly int distance;

        public static readonly Interval Semitone = new Interval(1);
        public static readonly Interval HalfStep = new Interval(1);
        public static readonly Interval MinorSecond = new Interval(1);

        public static readonly Interval Tone = new Interval(2);
        public static readonly Interval Step = new Interval(2);

        public static readonly Interval MinorThird = new Interval(3);
        public static readonly Interval MajorThird = new Interval(4);

        public static readonly Interval Fourth = new Interval(5);

        public static readonly Interval Tritone = new Interval(6);

        public static readonly Interval Fifth = new Interval(7);

        public static readonly Interval MinorSixth = new Interval(8);
        public static readonly Interval MajorSixth = new Interval(9);
        public static readonly Interval MinorSeventh = new Interval(10);
        public static readonly Interval MajorSeventh = new Interval(11);

        public static readonly Interval Octave = new Interval(12);

        public static readonly Interval MinorNinth = new Interval(13);
        public static readonly Interval Ninth = new Interval(14);
        

		public Interval(int semitones)
		{
			this.distance = semitones;
		}

		public int Semitones { get { return this.distance; } }

        public bool Ascending
        {
            get
            {
                return this.distance >= 0;
            }
        }

        public bool Descending
        {
            get
            {
                return this.distance < 0;
            }
        }

        public override int GetHashCode()
		{
			return this.distance.GetHashCode();
		}

		public override string ToString()
		{
            return IntervalNamer.NameOf(this);
		}

		public override bool Equals(object obj)
		{
			if (object.ReferenceEquals(obj, null)) return false;

			if (object.ReferenceEquals(this, obj)) return true;

			if (this.GetType() != obj.GetType()) return false;

			return this.Equals(obj as Interval);
		}

		public bool Equals(Interval other)
		{
			if (object.ReferenceEquals(other, null)) return false;

			return this.distance == other.distance;
		}

		public int CompareTo(object obj)
		{
			Interval other = obj as Interval;

			if (object.ReferenceEquals(other, null)) throw new ArgumentException("Argument is not a valid pitch value", "obj");

			return this.CompareTo(other);
		}

		public int CompareTo(Interval other)
		{
			if (object.ReferenceEquals(other, null)) throw new ArgumentNullException("other");

			return this.distance.CompareTo(other.distance);
		}

        public bool RelatedTo(Interval other)
        {
            if (object.ReferenceEquals(other, null)) throw new ArgumentNullException("other");

            return (this.distance % 12 == other.distance % 12);
        }

        public static bool operator ==(Interval left, Interval right)
		{
			if (object.ReferenceEquals(left, null)) return false;
			if (object.ReferenceEquals(right, null)) return false;

			return left.distance == right.distance;
		}

		public static bool operator !=(Interval left, Interval right)
		{
			if (object.ReferenceEquals(left, null)) return true;
			if (object.ReferenceEquals(right, null)) return true;

			return !(left == right);
		}

		public static bool operator <(Interval left, Interval right)
		{
			if (object.ReferenceEquals(left, null)) return false;
			if (object.ReferenceEquals(right, null)) return false;

			return left.CompareTo(right) < 0;
		}

		public static bool operator <=(Interval left, Interval right)
		{
			if (object.ReferenceEquals(left, null)) return false;
			if (object.ReferenceEquals(right, null)) return false;

			return left.CompareTo(right) <= 0;
		}

		public static bool operator >(Interval left, Interval right)
		{
			if (object.ReferenceEquals(left, null)) return false;
			if (object.ReferenceEquals(right, null)) return false;

			return left.CompareTo(right) > 0;
		}

		public static bool operator >=(Interval left, Interval right)
		{
			if (object.ReferenceEquals(left, null)) return false;
			if (object.ReferenceEquals(right, null)) return false;

			return left.CompareTo(right) >= 0;
		}

        public Solfege AsSolfege()
        {
            Solfege[] solfa = new Solfege[] {
                Solfege.Do,
                Solfege.Di,
                Solfege.Re,
                Solfege.Ri,
                Solfege.Mi,
                Solfege.Fa,
                Solfege.Fi,
                Solfege.So,
                Solfege.Si,
                Solfege.La,
                Solfege.Li,
                Solfege.Ti
            };

            int degree = this.distance % solfa.Length;

            return solfa[degree];
        }
	}

}

File: src/tedesco/interval.py
from __future__ import annotations
from dataclasses import dataclass
from functools import total_ordering
from typing import ClassVar

_NAME_TO_SEMITONES = {
    "unison": 0, 
    "minor second": 1, 
    "major second": 2, 
    "minor third": 3, 
    "major third": 4, 
    "perfect fourth": 5, 
    "diminished fifth": 6, 
    "perfect fifth": 7, 
    "minor sixth": 8, 
    "major sixth": 9, 
    "minor seventh": 10, 
    "major seventh": 11, 
    "octave": 12, 
    "minor ninth": 13, 
    "ninth": 14,
}

@total_ordering
@dataclass(frozen=True)
class Interval:
    """A chromatic interval measured in semitones (can be negative)."""
    semitones: int
    
    @classmethod
    def from_name(cls, name: str) -> "Interval":
        return cls(_NAME_TO_SEMITONES[name.strip().lower()])
        
    """Well-known values"""
    Unison: ClassVar["Interval"]
    HalfStep: ClassVar["Interval"]
    MinorSecond: ClassVar["Interval"]
    WholeStep: ClassVar["Interval"]
    MajorSecond: ClassVar["Interval"]
    MinorThird: ClassVar["Interval"]
    MajorThird: ClassVar["Interval"]
    PerfectFourth: ClassVar["Interval"]
    Tritone: ClassVar["Interval"]
    Fifth: ClassVar["Interval"]
    MinorSixth: ClassVar["Interval"]
    MajorSixth: ClassVar["Interval"]
    MinorSeventh: ClassVar["Interval"]
    MajorSeventh: ClassVar["Interval"]
    Octave: ClassVar["Interval"]
    MinorNinth: ClassVar["Interval"]
    Ninth: ClassVar["Interval"]

    def __repr__(self) -> str:
        sign = "+" if self.semitones >= 0 else "-"
        return f"<Interval {sign}{abs(self.semitones)} st>"
    
    # ---- comparisons required by @total_ordering ----
    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Interval):
            return NotImplemented
        return self.semitones == other.semitones

    def __lt__(self, other: "Interval") -> bool:
        if not isinstance(other, Interval):
            return NotImplemented
        return self.semitones < other.semitones
    
    def __add__(self, other):
        try:
            from .note import Note  # Lazy imported here to avoid circular import
        except:
            Note = None
        if (Note is not None and isinstance(other, Note)):
            return other + self
        return NotImplemented

    def __mul__(self, other):
        raise TypeError("Interval objects do not support multiplication")
    
    def __rmul__(self, other):
        raise TypeError("Interval objects do not support multiplication")
    
    def __truediv__(self, other):
        raise TypeError("Interval objects do not support division")
    
    def __floordiv__(self, other):
        raise TypeError("Interval objects do not support division")
    
    def __mod__(self, other):
        raise TypeError("Interval objects do not support modulo")

    def relatedTo(self, other: "Interval") -> bool:
        if not isinstance(other, Interval):
            return NotImplemented

        return self.semitones % 12 == other.semitones % 12

    def asSolfege(self) -> str:
        solfa = ["do", "di", "re", "ri", "mi", "fa", "fi", "so", "si", "la", "li", "ti"]

        degree = self.semitones % len(solfa)
        return solfa[degree]

    def name(self) -> str:
        names = [ "unison", "minor second", "major second", "minor third", "major third", "perfect fourth", "diminished fifth", 
                 "perfect fifth", "minor sixth", "major sixth", "minor seventh", "major seventh", "octave", "minor ninth", "ninth" ]
        
        distance = abs(self.semitones)

        if distance >= len(names):
            return "unknown"
            
        return names[distance]

        
# Define the static singletons *after* the class is created.
Interval.Unison         = Interval(0)
Interval.HalfStep       = Interval(1)
Interval.MinorSecond    = Interval(1)
Interval.MajorSecond    = Interval(2)
Interval.WholeStep      = Interval(2)
Interval.MinorThird     = Interval(3)
Interval.MajorThird     = Interval(4)
Interval.PerfectFourth  = Interval(5)
Interval.Tritone        = Interval(6)
Interval.PerfectFifth   = Interval(7)  
Interval.MinorSixth     = Interval(8)
Interval.MajorSixth     = Interval(9)
Interval.MinorSeventh   = Interval(10)
Interval.MajorSeventh   = Interval(11)
Interval.Octave         = Interval(12)
Interval.MinorNinth     = Interval(13)
Interval.Ninth          = Interval(14)

